File: internal/aggregate/document.go
package aggregate

import "mcp.chic.md/internal/moysklad"

// DocumentSummary is the compact form of a document list row. Sums in major units.
type DocumentSummary struct {
	ID           string  `json:"id"`
	Name         string  `json:"name"`
	Moment       string  `json:"moment"`
	Sum          float64 `json:"sum"`
	Paid         float64 `json:"paid,omitempty"`
	Counterparty string  `json:"counterparty,omitempty"`
	State        string  `json:"state,omitempty"`
	Store        string  `json:"store,omitempty"`
	Channel      string  `json:"channel,omitempty"`
	// Currency is the ISO code the amounts on THIS document are denominated in.
	// Empty means the account base currency. When set and different from the
	// account base, Sum/Paid are in this currency — do not mix with base-currency
	// totals. Only populated when the request expands rate.currency.
	Currency   string `json:"currency,omitempty"`
	Applicable bool   `json:"applicable"`
}

func DocumentSummaryOf(d moysklad.Document) DocumentSummary {
	s := DocumentSummary{
		ID:         d.ID,
		Name:       d.Name,
		Moment:     d.Moment,
		Sum:        MinorToMajor(d.Sum),
		Paid:       MinorToMajor(d.PayedSum),
		Applicable: d.Applicable,
	}
	if d.Rate != nil && d.Rate.Currency != nil {
		s.Currency = d.Rate.Currency.ISOCode
	}
	if d.Agent != nil {
		s.Counterparty = d.Agent.Name
	}
	if d.State != nil {
		s.State = d.State.Name
	}
	if d.Store != nil {
		s.Store = d.Store.Name
	}
	if d.SalesChannel != nil {
		s.Channel = d.SalesChannel.Name
	}
	return s
}

func DocumentSummaries(docs []moysklad.Document) []DocumentSummary {
	out := make([]DocumentSummary, 0, len(docs))
	for _, d := range docs {
		out = append(out, DocumentSummaryOf(d))
	}
	return out
}

// DocumentTotals are the grand totals across every matching document — so the
// model can answer "total sales for the period" without summing rows itself.
type DocumentTotals struct {
	Sum  float64 `json:"sum"`
	Paid float64 `json:"paid"`
}

// DocumentReport summarizes documents, totals over all of them, and truncates
// the detail list to limit. Order is preserved (the API sorts by moment desc),
// so Rows is the most recent documents while Totals covers the whole period.
func DocumentReport(docs []moysklad.Document, limit int) Report[DocumentSummary, DocumentTotals] {
	rows := DocumentSummaries(docs)
	var t DocumentTotals
	for _, r := range rows {
		t.Sum += r.Sum
		t.Paid += r.Paid
	}
	t.Sum = round2(t.Sum)
	t.Paid = round2(t.Paid)
	return newReport(rows, t, limit)
}

// DocumentDetail is a document with its expanded positions.
type DocumentDetail struct {
	DocumentSummary
	Description string        `json:"description,omitempty"`
	VatSum      float64       `json:"vatSum,omitempty"`
	DueDate     string        `json:"paymentDueDate,omitempty"`
	Positions   []PositionRow `json:"positions,omitempty"`
}

type PositionRow struct {
	Name     string  `json:"name"`
	Code     string  `json:"code,omitempty"`
	Quantity float64 `json:"quantity"`
	Price    float64 `json:"price"`
	Discount float64 `json:"discount,omitempty"` // percent
	Vat      int     `json:"vat,omitempty"`      // VAT rate, percent
	Total    float64 `json:"total"`              // price*qty net of discount
}

func DocumentDetailOf(d moysklad.Document) DocumentDetail {
	det := DocumentDetail{
		DocumentSummary: DocumentSummaryOf(d),
		Description:     d.Description,
		VatSum:          MinorToMajor(d.VatSum),
		DueDate:         d.PaymentPlannedMoment,
	}
	if d.Positions != nil {
		for _, p := range d.Positions.Rows {
			price := MinorToMajor(p.Price)
			// МойСклад's document Sum is already net of line discounts, so the
			// line total must subtract the discount too or the positions won't
			// reconcile with the header.
			total := price * p.Quantity * (1 - p.Discount/100)
			det.Positions = append(det.Positions, PositionRow{
				Name:     p.Assortment.Name,
				Code:     p.Assortment.Code,
				Quantity: p.Quantity,
				Price:    price,
				Discount: p.Discount,
				Vat:      p.Vat,
				Total:    round2(total),
			})
		}
	}
	return det
}

File: chic/aggregate/document.py
"""Document summaries and expanded detail (minor→major units)."""

from __future__ import annotations

from typing import Any

from chic.aggregate.models import (
    DocumentDetail,
    DocumentSummary,
    DocumentTotals,
    PositionRow,
    Report,
)
from chic.aggregate.money import minor_to_major, round2
from chic.aggregate.report import _truncate
from chic.moysklad.models import Document


def _attr_value(value: Any) -> Any:
    """Flatten a custom-attribute value to a display scalar.

    Custom-entity / linked attributes arrive as objects carrying their own
    ``name``; simple attributes are scalars already.
    """
    if isinstance(value, dict):
        return value.get("name") or value.get("value") or ""
    return value


def _attributes(d: Document) -> dict[str, Any]:
    return {a.name: _attr_value(a.value) for a in d.attributes if a.name}


def document_summary_of(d: Document) -> DocumentSummary:
    return DocumentSummary(
        id=d.id,
        name=d.name,
        moment=d.moment,
        sum=minor_to_major(d.sum),
        paid=minor_to_major(d.payed_sum),
        applicable=d.applicable,
        currency=(
            d.rate.currency.iso_code if d.rate is not None and d.rate.currency is not None else ""
        ),
        counterparty=d.agent.name if d.agent is not None else "",
        state=d.state.name if d.state is not None else "",
        store=d.store.name if d.store is not None else "",
        channel=d.sales_channel.name if d.sales_channel is not None else "",
    )


def document_summaries(docs: list[Document]) -> list[DocumentSummary]:
    return [document_summary_of(d) for d in docs]


def document_report(docs: list[Document], limit: int) -> Report[DocumentSummary, DocumentTotals]:
    rows = document_summaries(docs)
    totals = DocumentTotals(
        sum=round2(sum(r.sum for r in rows)),
        paid=round2(sum(r.paid for r in rows)),
    )
    shown, full, truncated = _truncate(rows, limit)
    return Report[DocumentSummary, DocumentTotals](
        totals=totals, row_count=full, returned=len(shown), truncated=truncated, rows=shown
    )


def document_detail_of(d: Document) -> DocumentDetail:
    base = document_summary_of(d)
    positions: list[PositionRow] = []
    if d.positions is not None:
        for p in d.positions.rows:
            price = minor_to_major(p.price)
            # MoySklad's document Sum is already net of line discounts, so the
            # line total must subtract the discount too to reconcile with the header.
            total = price * p.quantity * (1 - p.discount / 100)
            positions.append(
                PositionRow(
                    name=p.assortment.name,
                    code=p.assortment.code,
                    quantity=p.quantity,
                    price=price,
                    discount=p.discount,
                    vat=p.vat,
                    total=round2(total),
                )
            )
    return DocumentDetail(
        **base.model_dump(),
        description=d.description,
        vat_sum=minor_to_major(d.vat_sum),
        payment_due_date=d.payment_planned_moment,
        positions=positions,
        attributes=_attributes(d),
    )
